File: elixir/proverb/lib/proverb.ex
defmodule Proverb do
  @doc """
  Generate a proverb from a list of strings.
  """
  @spec recite(strings :: [String.t()]) :: String.t()
  def recite(strings) do
    case strings do
      []          -> ""
      [word]      -> moral(word)
      [word | _]
        -> ( strings
             |> Enum.chunk_every(2, 1, :discard)
             |> Enum.map(fn [a, b] -> stanza(a, b) end)
             |> Enum.join()
           ) <> moral(word)
    end
  end

  defp moral(word) do
    "And all for the want of a #{word}.\n"
  end

  defp stanza(a, b) do
    "For want of a #{a} the #{b} was lost.\n"
  end
end

File: csharp/proverb/Proverb.cs
using System;

public static class Proverb
{
    public static string[] Recite(string[] subjects)
    {
        if (subjects == Array.Empty<string>()) return Array.Empty<string>();
        string[] result = new string[subjects.Length];
        for (int i = 0; i < subjects.Length - 1; i++)
        {
            result[i] = $"For want of a {subjects[i]} the {subjects[i+1]} was lost.";
        }
        result[^1] = $"And all for the want of a {subjects[0]}.";
        return result;
    }
}